### PR TITLE
feat: incresed MAX_NATIONAL_GENERATION_MW and dynamic yMax value

### LINKED ISSUE #559: Handle National Capacity over 12GW
## Describe the bug
When the forecast (or P90) is over 12GW for the National forecast, the chart has to stretch and set a new max Y value. 

## Expected behavior
This should either be handled explicitly, or set a higher National Capacity value.

## Additional context
Currently default national capacity is 12, actual is 17GW+

### PR DESCRIPTION
**Pull request by suvanbanerjee**
# Pull Request

## Description

Increased the value of MAX_NATIONAL_GENERATION_MW to 18000 (18GW) and to prevent graph from stretching when the max value is exceeded add a buffer of 1000 and round to next 1000th value and set it as yMax  

Fixes #559

## How Has This Been Tested?

Added jest Test cases to new util function
image

When yMax is 18000 
image

when yMax is 1000 to indicate data overflow.
image

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by suvanbanerjee** [comment]
@braddf @peterdudfield can you review this pr please

File: apps/nowcasting-app/constant.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@ export const API_PREFIX = process.env.NEXT_PUBLIC_API_PREFIX || "https://api-dev
 export const SITES_API_PREFIX =
   process.env.NEXT_PUBLIC_SITES_API_PREFIX || "https://api-site-dev.quartz.solar";
 export const MAX_POWER_GENERATED = 500;
-export const MAX_NATIONAL_GENERATION_MW = 12000;
+export const MAX_NATIONAL_GENERATION_MW = 18000;
```
Comment (braddf) [inline apps/nowcasting-app/constant.ts:6]: If we have the additional logic you've added for cases where we do go over this threshold, I think we can maybe keep the lower max value here so that the overall graph can still use a good amount of the height, and avoid startling the user into thinking our generation is (visually) much less than it was before – could we try it with the `12000` back in and see if it works for both below and above that (UK solar is very rarely going to be hitting these heights for the near future!)

File: apps/nowcasting-app/components/charts/pv-remix-chart.tsx [hunk 2]
```diff
@@ -67,6 +67,29 @@ const PvRemixChart: FC<{
...
+          .map(([_, value]) => value as number)
+      )
+    );
+
+    const buffer = 1000; // This buffer is added to the max value to ensure the chart is not too close to the top
+    const roundedMax = Math.ceil((maxDataValue + buffer) / 1000) * 1000;
+
+    return Math.max(MAX_NATIONAL_GENERATION_MW, roundedMax);
+  }, [chartData]);
+
```
Comment (braddf) [inline apps/nowcasting-app/components/charts/pv-remix-chart.tsx:92]: This is great, thank you! I was hoping whoever picked this up would put in some nice logic here ⭐ 
BUT that does mean that a function like this encourages me to be extra picky 😁 for both testing and understanding of what this function is doing e.g. by newcomers, it would be amazing if this could be:
- separated into a explicitly-typed, pure function (for input/output, nothing crazy);
- in a separate file, for neatness and testability;
- with a handful of unit tests to make sure it's doing exactly what we want it to do with <12k, ==12k, >12k etc.
- and just for personal preference, maybe exclude any weird numbers like 13k/17k because they won't divide nicely into ticks with round numbers! 😅 

Aware that this is more strict than a lot of the nearby code, but trying to keep best practices tight for any new code from here on out 🙏

**Review by braddf — CHANGES_REQUESTED** [review]
Nice, thanks @suvanbanerjee – left a couple of comments, let me know if anything is unclear 👍

File: apps/nowcasting-app/components/charts/pv-remix-chart.tsx (hunk 2 above)
Comment (suvanbanerjee) [reply apps/nowcasting-app/components/charts/pv-remix-chart.tsx:74]: @braddf 
- put the code in utils
- added test
- rounding to next 2000 to avoid 13k 17k...

File: apps/nowcasting-app/components/helpers/utils.ts [hunk 3]
```diff
@@ -573,3 +573,30 @@ export const createBucketObject: (
...
+        .filter(
+          ([key, value]) =>
+            typeof value === "number" && key !== "formattedDate" && !key.includes("TIME")
+        )
+        .map(([_, value]) => value as number)
+    )
+  );
+
+  const valueWithBuffer = maxDataValue + 1000;
+  const roundingFactor = 2000;
```
Comment (braddf) [inline apps/nowcasting-app/components/helpers/utils.ts:598]: Probably don't need to arbitrarily add/round to ~2000 above the max, we can just explicitly check for the `roundedMax` being in `[13000, 17000, 23000]` or something similar to keep the chart bounds closer to the data?

File: apps/nowcasting-app/components/charts/pv-remix-chart.tsx [hunk 4]
```diff
@@ -128,7 +110,7 @@ const PvRemixChart: FC<{
               timeOfInterest={selectedTime}
               setTimeOfInterest={setSelectedTime}
               data={chartData}
-              yMax={yMax}
+              yMax={Math.max(yMax, MAX_NATIONAL_GENERATION_MW)}
```
Comment (braddf) [inline apps/nowcasting-app/components/charts/pv-remix-chart.tsx:113]: Can we include this `Math.max` in the function itself, passing the `MAX` constant as a second param? Will be neater for testing ☺️

File: apps/nowcasting-app/components/helpers/utils.ts [hunk 5]
```diff
@@ -573,3 +573,30 @@ export const createBucketObject: (
     increment
   };
 };
+
+export function calculateChartYMax<T extends Record<string, any>>(
```
Comment (braddf) [inline apps/nowcasting-app/components/helpers/utils.ts:577]: Unless we refactor this to be used in multiple places, let's use the concrete Type for `ChartData` – it might not be set yet so we can handle that, but if it is, it will then be the correct shape/have the right properties for what we need and the tests will be more specific 👍

File: apps/nowcasting-app/components/helpers/utils.test.ts [hunk 6]
```diff
@@ -181,3 +182,74 @@ describe("prettyPrintDayLabelWithDate", () => {
...
+    const chartData = [
+      { value1: 5000, value2: 8000 },
+      { value1: 3000, value2: 7000 }
+    ];
+    const result = utils.calculateChartYMax(chartData);
+    expect(result).toBe(10000); // (8000 + 1000) rounded up to nearest 2000
+  });
+
+  test("should ignore formattedDate fields", () => {
+    const chartData = [{ value: 6000, formattedDate: 15000 }];
```
Comment (braddf) [inline apps/nowcasting-app/components/helpers/utils.test.ts:218]: Nice, good to check this – this should be caught if you implement the concrete Type, but formattedDate I think is a `string`

File: apps/nowcasting-app/components/helpers/utils.ts [hunk 7]
```diff
@@ -573,3 +573,30 @@ export const createBucketObject: (
...
+    return 0;
+  }
+
+  const maxDataValue = Math.max(
+    0,
+    ...chartData.flatMap((dataPoint) =>
+      Object.entries(dataPoint)
+        .filter(
+          ([key, value]) =>
+            typeof value === "number" && key !== "formattedDate" && !key.includes("TIME")
```
Comment (braddf) [inline apps/nowcasting-app/components/helpers/utils.ts:591]: Not sure we have any `TIME` fields in this Type?

File: apps/nowcasting-app/components/helpers/utils.test.ts [hunk 8]
```diff
@@ -181,3 +182,74 @@ describe("prettyPrintDayLabelWithDate", () => {
...
+    const result = utils.calculateChartYMax(chartData);
+    expect(result).toBe(8000); // (6000 + 1000) rounded up to nearest 2000
+  });
+
+  test("should ignore fields containing TIME", () => {
```
Comment (braddf) [inline apps/nowcasting-app/components/helpers/utils.test.ts:223]: Remove if no `TIME` fields as per other comment

File: apps/nowcasting-app/components/helpers/utils.test.ts [hunk 9]
```diff
@@ -181,3 +182,74 @@ describe("prettyPrintDayLabelWithDate", () => {
...
+    expect(result).toBe(6000); // (4500 + 1000) rounded up to nearest 2000
+  });
+
+  test("should handle edge case where values exactly match rounding threshold", () => {
+    const chartData = [{ value: 5000 }];
+    const result = utils.calculateChartYMax(chartData);
+    expect(result).toBe(8000); // (5000 + 1000) rounded up to nearest 2000
+  });
+
+  test("should consider yMax if value is greater", () => {
```
Comment (braddf) [inline apps/nowcasting-app/components/helpers/utils.test.ts:235]: With this test and the below, make sure it's clear what "value" we are referring to and therefore which thing is greater/less – took me a couple of reads to confirm!

Comment (braddf) [reply apps/nowcasting-app/components/helpers/utils.test.ts:235]: And maybe "return" instead of "consider", we're considering a lot of things here! 😄

**Review by braddf — CHANGES_REQUESTED** [review]

File: apps/nowcasting-app/components/charts/pv-remix-chart.tsx (hunk 4 above)
Comment (suvanbanerjee) [reply apps/nowcasting-app/components/charts/pv-remix-chart.tsx:113]: I tried it earlier but got the result as below 
image

where its streched

Comment (suvanbanerjee) [reply apps/nowcasting-app/components/charts/pv-remix-chart.tsx:113]: ig this issue is when values are odd, if I apply the 2000 rounding its working

File: apps/nowcasting-app/components/helpers/utils.ts (hunk 7 above)
Comment (suvanbanerjee) [reply apps/nowcasting-app/components/helpers/utils.ts:591]: Opps it is GENERATION will remove this case

File: apps/nowcasting-app/components/helpers/utils.ts (hunk 3 above)
Comment (suvanbanerjee) [reply apps/nowcasting-app/components/helpers/utils.ts:598]: When the roundedMax is odd the stretching issue is occurring, hence I think we should keep the 2000 rounding

**Comment by suvanbanerjee** [comment]
@braddf bumping this up

File: apps/nowcasting-app/components/helpers/utils.ts (hunk 3 above)
Comment (braddf) [reply apps/nowcasting-app/components/helpers/utils.ts:598]: Hey Suvan, I think I left a reply to you on here but can't see your previous comment, maybe you deleted it?
Sounds like it might need to have the Y-axis ticks set explicitly with a function then in that case, which you should see in other charts – I think in the `gsp-pv-remix-chart` maybe from memory? 

For context, the main design point here is to only change the Y Max bound when we need to, i.e. when it's actually going to be more than 12000, and otherwise users can assume that the axis is fixed, and that each day they can mentally compare the heights and shapes to previous days they view the UI.

Comment (suvanbanerjee) [reply apps/nowcasting-app/components/helpers/utils.ts:598]: @braddf now I got the clarity thanks :D pushing patch in 30min

**Comment by suvanbanerjee** [comment]
@braddf made the changes please review

File: apps/nowcasting-app/components/helpers/chartUtils.ts [hunk 10]
```diff
@@ -62,6 +62,8 @@ export const getTicks = (yMax: number, yMax_levels: number[]) => {
   const fifth = yMax / 5;
   const seventh = yMax / 7;
   const testTicksToAdd = (fractionN: number) => {
+    if (!Number.isFinite(fractionN) || fractionN <= 0) return;
```
Comment (braddf) [inline apps/nowcasting-app/components/helpers/chartUtils.ts:65]: Was this just for extra safety? Checking for a divide by zero or a NaN _should_ be unnecessary here with Typescript, but technically shouldn't do any harm; the `<= 0` actually excludes the smaller values needed for lower capacity GSP regions – this is my fault for not having written a test here to catch this, but this function needs to work for everything (within reason, they won't all look perfect, especially those pesky surprise primes) from 1 to 12000 as a `yMax` value! Hence, the line below rounding the value also doesn't work, e.g. if the GSP capacity is 2MW, need some non-round ticks:

image

File: apps/nowcasting-app/components/helpers/utils.test.ts [hunk 11]
```diff
@@ -181,3 +182,59 @@ describe("prettyPrintDayLabelWithDate", () => {
...
+
+  test("should round up to nearest 2000 with buffer of 1000", () => {
+    const chartData = [{ value: 12500 }];
+    const result = utils.calculateChartYMax(chartData);
+    expect(result).toBe(14000); // 12500 + 1000 = 13500 rounded up to nearest 2000, ie 14000
+  });
+
+  test("should handle multiple numeric values", () => {
+    const chartData = [
+      { value1: 5000, value2: 8000 },
```
Comment (braddf) [inline apps/nowcasting-app/components/helpers/utils.test.ts:210]: With the `ChartData` type in the function as per other comment below, this needs to fit that type too

File: apps/nowcasting-app/components/helpers/utils.ts [hunk 12]
```diff
@@ -573,3 +573,30 @@ export const createBucketObject: (
     increment
   };
 };
+export function calculateChartYMax(
+  chartData: Array<{ [key: string]: number | string | number[] }> | null | undefined,
```
Comment (braddf) [inline apps/nowcasting-app/components/helpers/utils.ts:577]: Please set this `chartData` param to the `ChartData` type so we can enforce the type properly 👍

File: apps/nowcasting-app/components/charts/pv-remix-chart.tsx [hunk 13]
```diff
@@ -68,26 +70,7 @@ const PvRemixChart: FC<{
...
-          )
-          .map(([_, value]) => value as number)
-      )
-    );
-
-    const buffer = 1000; // This buffer is added to the max value to ensure the chart is not too close to the top
-    const roundedMax = Math.ceil((maxDataValue + buffer) / 1000) * 1000;
-
-    return Math.max(MAX_NATIONAL_GENERATION_MW, roundedMax);
+    return calculateChartYMax(chartData, MAX_NATIONAL_GENERATION_MW);
```
Comment (braddf) [inline apps/nowcasting-app/components/charts/pv-remix-chart.tsx:73]: Nice 👌

**Review by braddf — CHANGES_REQUESTED** [review]
Couple of things but nearly there!

File: apps/nowcasting-app/components/helpers/chartUtils.ts (hunk 10 above)
Comment (suvanbanerjee) [reply apps/nowcasting-app/components/helpers/chartUtils.ts:65]: because of the long loops during testing i add this adjustment to ensure the tests run correctly. 
I'll update the test cases for value less than 2MW as well! :D

**Comment by suvanbanerjee** [comment]
@braddf made the changes :D let's have a last round of review 🤞

**Review by braddf — APPROVED** [review]
Great stuff! Thanks @suvanbanerjee – excellent debut to the repo 🙌
